### PR TITLE
feat: implement `ddev pull upsun` provider integration, fixes #5446

### LINKED ISSUE #5446: Add `ddev pull upsun` support
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

Platform.sh has a new product called `Upsun`, which needs `ddev pull upsun`

### Describe your solution

* Add the upsun cli into the web container
* Add upsun.yaml
* Add CI for testing it

### Describe alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stasadev**
## The Issue

- #5446

## How This PR Solves The Issue

Adds integration for [Upsun](https://upsun.com/).

## Manual Testing Instructions

Docs are here https://ddev--5529.org.readthedocs.build/en/5529/users/providers/upsun/

Try `ddev pull upsun` in Upsun project
Try `ddev push upsun`

## Automated Testing Overview

`TestUpsunPull` and `TestUpsunPush`.

### CODE REVIEW

**Comment by rfay** [comment]
ddev-windows-mutagen isn't supposed to run on PRs...  I guess it's because you did this PR on ddev org. And of course it's irrelevant to this PR.  But TestUpsunPull/Push worked, yay, https://github.com/ddev/ddev/actions/runs/6829336445/job/18575278444?pr=5529

File: docs/content/users/providers/upsun.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+3. Run [`ddev restart`](../usage/commands.md#restart).
+4. Run `ddev pull upsun`. After you agree to the prompt, the current upstream databases and files will be downloaded.
+5. Optionally use `ddev push upsun` to push local files and database to Upsun. The [`ddev push`](../usage/commands.md#push) command can potentially damage your production site, so we don’t recommend using it.
+
+### Managing Multiple Databases
+
+If your project has only one database, it will automatically be pulled into and pushed from DDEV’s `'db'` database.
+
+If your project has multiple databases, they’ll all be pulled into DDEV with their respective remote names. You can optionally designate a *primary* to use DDEV’s default `'db'` database, which may be useful in some cases—particularly if you’ve been using the default solo-database behavior and happened to add another one to your project.
```
[the remark this answers is not in the document]
Comment (stasadev) [reply docs/content/users/providers/upsun.md:46]: I may have missed this because I was only using one database.

**Comment by rfay** [comment]
Multiple databases does not yet work, because `upsun relationships` has a different structure than `platform relationships`. 

To configure multiple databases in upsun you need to have them configured in the .upsun/config.yaml:

```
services:
  mariadb:
    type: mariadb:10.6
    configuration:
      schemas:
        - main
        - junk
      endpoints:
        mysql:
          default_schema: main
          privileges:
            main: admin
            junk: admin
```


`upsun relationships` looks like this:
```
mariadb:
    -
        host: mariadb.internal
        hostname: 2jins5mtttge2mfrfdcxkxkpte.mariadb.service._.ca-1.platformsh.site
        cluster: 47kohks3azo5c-main-bvxea6i
        service: mariadb
        rel: mysql
        scheme: mysql
        username: mysql
        password: 365d3d4b06cdf5173da0bf86008de9bf
        port: 3306
        epoch: 0
        path: main
        query:
            is_master: true
        fragment: null
        public: false
        host_mapped: false
        type: 'mariadb:10.6'
        ip: 169.254.156.174
        url: 'mysql://mysql:365d3d4b06cdf5173da0bf86008de9bf@mariadb.internal:3306/main'
```

But `platform relationships` lists the databases (path):

```
database:
    -
        username: user
        fragment: null
        ip: 169.254.193.18
        cluster: 5bviezdszcmrg-platform-pull-7tsp6cq
        host: database.internal
        path: main
        query:
            is_master: true
        password: ‘’
        port: 3306
        host_mapped: false
        service: db
        hostname: jzglbn36ugubmwptswhuwv62sm.db.service._.ca-1.platformsh.site
        epoch: 0
        rel: mysql
        scheme: mysql
        type: ‘mariadb:10.4’
        public: false
        url: ‘mysql://user:@database.internal:3306/main’
```

I guess the difference is the key, `database` as opposed to `mariadb` ?

**Comment by rfay** [comment]
I guess the logic in https://github.com/ddev/ddev/blob/a0e022ec9a50fd6de204a07d8c60d2c75b532001/pkg/ddevapp/dotddev_assets/providers/platform.yaml#L54-L57 is inadequate for upsun.

It's worth a little effort to fix it, or we can just say "only one db for now"

**Comment by stasadev** [comment]
I will look into it tomorrow, and if I can't figure it out quickly, we can proceed with "only one db for now".

**Comment by rfay** [comment]
So far I haven't found a good way to list the databases provided. Have questions going in slack. You don't need to fiddle with it until they reply. thanks!

**Comment by rfay** [comment]
Rebased, but didn't re-push. This will need to be re-pushed before it's merged.

**Review by rfay — APPROVED** [review]
Looks great to me. I manually tested pull with rfay-d10, and branched a new `testpush` environment and did `ddev push upsun --environment=UPSUN_PROJECT=testpush` and it worked out fine.

**Comment by rfay** [comment]
I updated the ddev-webserver image.

**Pull request merged**